File: src/core/engine/library/arrays.ts
/**
 * Arrays, `stdlib.md` section 12's list of calls and `language.md` 5.2's value.
 *
 * Only two things an array does are instructions: the literal, which is
 * `ARRAY`, and the subscript, which is `ELEM`. Everything else is a call, and
 * the reason is in `compiled-program.md` 4.9: an instruction per array
 * operation would be twenty opcodes that a library function already expresses,
 * and the instruction set is deliberately dull.
 *
 * **An index outside an array is an error, not absence.** That is the opposite
 * of a history read past the start of the dataset, and the difference is the
 * point: an array has an extent the script chose, so an index outside it is a
 * mistake rather than a missing measurement.
 *
 * **Sorting is a total order written down here**, because a comparison that
 * fell back to the host language's would sort strings by a locale and mixed
 * contents by whatever the sort happened to compare first, and two engines
 * would then disagree about the contents of an array a script can read back.
 */
import type { ArrayObject } from '../values/index.js';
import type { Value } from '../values/index.js';
import { isNumber, reference, valuesEqual } from '../values/index.js';
import { entry, refAt, stringAt, valueAt, wholeAt } from './binding.js';
import type { CallContext, ManifestEntry } from './binding.js';
import { safe } from './state.js';

/** The array a handle names, or nothing when the handle is absent or stale. */
function held(ctx: CallContext, args: readonly Value[], index: number): ArrayObject | undefined {
  const handle = refAt(args, index);
  if (handle === null) return undefined;
  const object = ctx.heap.get(handle.id);
  return object !== undefined && object.kind === 'array' ? object : undefined;
}

/** Announces a change before making it, so the bar's undo journal has the old one. */
function changing(ctx: CallContext, args: readonly Value[], index: number): ArrayObject | undefined {
  const handle = refAt(args, index);
  if (handle === null) return undefined;
  const object = ctx.heap.get(handle.id);
  if (object === undefined || object.kind !== 'array') return undefined;
  ctx.heap.touch(handle.id);
  return object;
}

function grew(ctx: CallContext, array: ArrayObject, args: readonly Value[], by: number): void {
  ctx.heap.countElements(by);
  ctx.guard.array(ctx.span, ctx.nameOf(valueAt(args, 0)), array.items.length);
}

/** The index an array call was given, checked against the array's own extent. */
function within(
  ctx: CallContext,
  args: readonly Value[],
  at: number,
  array: ArrayObject,
  limit: number,
): number {
  const index = args[at];
  if (index === undefined || index === null || !isNumber(index) || !Number.isInteger(index) ||
      index < 0 || index >= limit) {
    ctx.guard.badIndex(
      ctx.span,
      ctx.nameOf(valueAt(args, 0)),
      index === undefined || index === null ? 'none' : String(index),
      array.items.length,
    );
  }
  return index;
}

/** The order `sort` uses, written down so two engines cannot differ. */
function rank(a: Value, b: Value): number {
  if (a === null) return b === null ? 0 : 1;
  if (b === null) return -1;
  if (typeof a === 'number' && typeof b === 'number') return a < b ? -1 : a > b ? 1 : 0;
  if (typeof a === 'string' && typeof b === 'string') return a < b ? -1 : a > b ? 1 : 0;
  if (typeof a === 'boolean' && typeof b === 'boolean') return a === b ? 0 : a ? 1 : -1;
  return 0;
}

function numbersOf(array: ArrayObject): (number | null)[] {
  return array.items.map((item) => (isNumber(item) ? item : null));
}

/**
 * The five reductions over an array's contents.
 *
 * These are the array forms of names the numeric library also carries over a
 * series, and they are written here rather than called through because the
 * library's forms take a length and a lookback and these take an extent the
 * script chose. The accumulation order is the same one: index order, oldest
 * first, summed fresh.
 *
 * An absent element makes the answer absent rather than being skipped, on the
 * same rule as every lookback: a mean over an array with a hole in it is a mean
 * over fewer values than the script thinks it has. `sum` of an empty array is
 * zero, and `avg`, `min`, `max` and `stdev` of one are absent, because there is
 * a sum of nothing and there is no mean of nothing.
 */
function reduce(array: ArrayObject): (number | null)[] | null {
  const values = numbersOf(array);
  for (const value of values) if (value === null) return null;
  return values;
}

function sumOf(array: ArrayObject): number | null {
  const values = reduce(array);
  if (values === null) return null;
  let total = 0;
  for (const value of values) total += value as number;
  return safe(total);
}

function extremeOf(array: ArrayObject, wantHigh: boolean): number | null {
  const values = reduce(array);
  if (values === null || values.length === 0) return null;
  let best = values[0] as number;
  for (const value of values) {
    const one = value as number;
    if (wantHigh ? one > best : one < best) best = one;
  }
  return safe(best);
}

function meanOf(array: ArrayObject): number | null {
  const values = reduce(array);
  if (values === null || values.length === 0) return null;
  const total = sumOf(array);
  return total === null ? null : safe(total / values.length);
}

function spreadOf(array: ArrayObject): number | null {
  const values = reduce(array);
  const mean = meanOf(array);
  if (values === null || mean === null) return null;
  let squares = 0;
  for (const value of values) {
    const deviation = (value as number) - mean;
    squares += deviation * deviation;
  }
  const variance = safe(squares / values.length);
  return variance === null || variance < 0 ? null : safe(Math.sqrt(variance));
}

function newArray(ctx: CallContext, items: Value[]): Value {
  ctx.guard.array(ctx.span, 'the array', items.length);
  return reference(ctx.heap.allocate({ kind: 'array', items }));
}

export const ARRAY_ENTRIES: readonly ManifestEntry[] = [
  entry('size', 'arr', (ctx, args) => held(ctx, args, 0)?.items.length ?? null),

  entry('element', 'arr i', (ctx, args) => {
    const array = held(ctx, args, 0);
    if (array === undefined) return null;
    return array.items[within(ctx, args, 1, array, array.items.length)] ?? null;
  }),

  entry('set', 'arr i v', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    array.items[within(ctx, args, 1, array, array.items.length)] = valueAt(args, 2);
    return null;
  }),

  entry('push', 'arr v', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    array.items.push(valueAt(args, 1));
    grew(ctx, array, args, 1);
    return null;
  }),

  entry('pop', 'arr', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    if (array.items.length === 0) {
      ctx.guard.badIndex(ctx.span, ctx.nameOf(valueAt(args, 0)), 'the last element', 0);
    }
    const value = array.items.pop() ?? null;
    ctx.heap.countElements(-1);
    return value;
  }),

  entry('shift', 'arr', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    if (array.items.length === 0) {
      ctx.guard.badIndex(ctx.span, ctx.nameOf(valueAt(args, 0)), 'the first element', 0);
    }
    const value = array.items.shift() ?? null;
    ctx.heap.countElements(-1);
    return value;
  }),

  entry('unshift', 'arr v', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    array.items.unshift(valueAt(args, 1));
    grew(ctx, array, args, 1);
    return null;
  }),

  // Insert accepts the position one past the end, which is an append: a loop
  // that inserts at `size(arr)` is how a script builds a list in order, and
  // refusing it would make the last element a special case.
  entry('insert', 'arr i v', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    const at = within(ctx, args, 1, array, array.items.length + 1);
    array.items.splice(at, 0, valueAt(args, 2));
    grew(ctx, array, args, 1);
    return null;
  }),

  entry('remove', 'arr i', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    const at = within(ctx, args, 1, array, array.items.length);
    const removed = array.items.splice(at, 1)[0] ?? null;
    ctx.heap.countElements(-1);
    return removed;
  }),

  entry('slice', 'arr from to', (ctx, args) => {
    const array = held(ctx, args, 0);
    if (array === undefined) return null;
    const from = wholeAt(ctx, 'slice', 'from', args, 1);
    const to = wholeAt(ctx, 'slice', 'to', args, 2);
    if (from === null || to === null) return null;
    return newArray(ctx, array.items.slice(from, to));
  }),

  entry('copy', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : newArray(ctx, [...array.items]);
  }),

  entry('indexOf', 'arr v', (ctx, args) => {
    const array = held(ctx, args, 0);
    if (array === undefined) return null;
    const wanted = valueAt(args, 1);
    for (let i = 0; i < array.items.length; i += 1) {
      if (valuesEqual(array.items[i] ?? null, wanted)) return i;
    }
    return -1;
  }),

  entry('arrayEqual', 'a b', (ctx, args) => {
    const a = held(ctx, args, 0);
    const b = held(ctx, args, 1);
    if (a === undefined || b === undefined) return null;
    if (a.items.length !== b.items.length) return false;
    for (let i = 0; i < a.items.length; i += 1) {
      if (!valuesEqual(a.items[i] ?? null, b.items[i] ?? null)) return false;
    }
    return true;
  }),

  entry('sort', 'arr order', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    const descending = stringAt(args, 1) === 'desc';
    array.items.sort((a, b) => (descending ? rank(b, a) : rank(a, b)));
    return null;
  }),

  entry('reverse', 'arr', (ctx, args) => {
    const array = changing(ctx, args, 0);
    if (array === undefined) return null;
    array.items.reverse();
    return null;
  }),

  entry('sum', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : sumOf(array);
  }),

  entry('avg', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : meanOf(array);
  }),

  entry('min', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : extremeOf(array, false);
  }),

  entry('max', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : extremeOf(array, true);
  }),

  entry('stdev', 'arr', (ctx, args) => {
    const array = held(ctx, args, 0);
    return array === undefined ? null : spreadOf(array);
  }),
];

/**
 * Emptying an array, which shares its name and its arity with emptying a grid.
 *
 * `stdlib.md` 14.3 gives `clear` two signatures, and both take one argument, so
 * the compiled program's `lib.functions` carries one entry for the pair and the
 * engine is the place that tells them apart. It does it by asking the heap what
 * the handle points at, which is the only run-time dispatch in the whole
 * library and is here because the format leaves it nowhere else to be.
 */
export function clearArray(ctx: CallContext, args: readonly Value[]): boolean {
  const array = changing(ctx, args, 0);
  if (array === undefined) return false;
  ctx.heap.countElements(-array.items.length);
  array.items.length = 0;
  return true;
}

File: engine/openscript/arrays.py
"""Arrays: the value of ``language.md`` 14.1 and the calls ``stdlib.md`` lists.

Two things an array does are instructions, the literal ``ARRAY`` and the
subscript ``ELEM``, and everything else is a call. The calls are here rather
than in ``library/``, because that package raises nothing and an array call has
two refusals of its own that no checker can make before the bar:

- **An index outside the array is OS4004**, not absence. That is the opposite of
  a history read past the start of the dataset, and the difference is the point:
  an array has an extent the script chose, so an index outside it is a mistake
  rather than a missing measurement. Taking an element from an empty array is
  the same refusal, naming the element that is not there.
- **An array past its element ceiling is OS5002.** The ceiling is a million
  elements and it is the language's own number, not a host's, so a script
  refused here is refused by every conforming engine.

Neither knows where in the source the call was written, which is the machine's
fact, so each is raised as ``Refused`` and the machine turns it into a
diagnostic at its own position (``Machine._positioned``).

**Every answer is written to agree with the first engine's to the bit.** The
reductions accumulate in index order, oldest first, as ``src/core/engine/
library/arrays.ts`` does; ``sort`` uses one total order written down below
rather than the host language's; and an absent element makes a reduction absent
rather than being skipped, because a mean over an array with a hole in it is a
mean over fewer values than the script thinks it has.

**Rollback needs nothing here.** An array is the object itself, held in a cell,
and the checkpoint is one deep copy over the cells, so a re-executed moving bar
starts from the array its checkpoint held.
"""

import functools
import math
from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple

from .values import ABSENT, ArrayValue, equal, finite, is_number, is_whole

#: ``language.md`` 14.1: an array holds at most this many elements.
ELEMENT_CEILING = 1_000_000

#: The name a refusal gives the array. The first engine names it from the
#: program's debug names, which this machine does not keep; no case compares it.
UNNAMED = "the array"


class Refused(Exception):
    """A refusal an array call makes, still to be given the call's position."""

    def __init__(self, code: str, **values: Any) -> None:
        super().__init__(code)
        self.code = code
        self.values = values


def literal(held: Sequence[Any]) -> ArrayValue:
    """``ARRAY``: a new array every time it executes, never a pooled constant."""
    return _new(list(held))


def element(array: Any, index: Any) -> Any:
    """``ELEM`` and ``element(arr, i)``: the element, or OS4004 outside the extent."""
    size = len(array.elements) if isinstance(array, ArrayValue) else 0
    if not isinstance(array, ArrayValue):
        raise Refused("OS4004", index=_spelled(index), name=UNNAMED, size=size)
    return array.elements[_within(array, index, size)]


def _new(items: List[Any]) -> ArrayValue:
    _grew(len(items))
    return ArrayValue(items)


def _grew(size: int) -> None:
    if size > ELEMENT_CEILING:
        raise Refused("OS5002", max=ELEMENT_CEILING, name=UNNAMED, size=size)


def _spelled(index: Any) -> str:
    return "none" if index is ABSENT else str(index)


def _within(array: ArrayValue, index: Any, limit: int) -> int:
    """The index a call was given, checked against the array's own extent."""
    if not is_whole(index) or not 0 <= index < limit:
        raise Refused("OS4004", index=_spelled(index), name=UNNAMED, size=len(array.elements))
    return int(index)


def _array(arguments: Sequence[Any], at: int = 0) -> Optional[ArrayValue]:
    held = arguments[at] if at < len(arguments) else ABSENT
    return held if isinstance(held, ArrayValue) else None


def _empty(array: ArrayValue, which: str) -> None:
    if not array.elements:
        raise Refused("OS4004", index=which, name=UNNAMED, size=0)


def _rank(a: Any, b: Any) -> int:
    """The order ``sort`` uses: absence last, then numbers, strings by code point, bools.

    Two values of different kinds compare equal, so a stable sort leaves them
    where they were, which is what the first engine's comparator does. Python's
    string order is code point order, which is ``language.md`` 9.3's.
    """
    if a is ABSENT:
        return 0 if b is ABSENT else 1
    if b is ABSENT:
        return -1
    both = (type(a), type(b))
    if is_number(a) and is_number(b) or both == (str, str):
        return -1 if a < b else 1 if a > b else 0
    if both == (bool, bool):
        return 0 if a == b else 1 if a else -1
    return 0


def _numbers(array: ArrayValue) -> Optional[List[float]]:
    """The elements as numbers, or nothing when any of them is not one."""
    out: List[float] = []
    for one in array.elements:
        if not is_number(one):
            return None
        out.append(float(one))
    return out


def _sum(array: ArrayValue) -> Any:
    values = _numbers(array)
    if values is None:
        return ABSENT
    total = 0.0
    for one in values:
        total += one
    return finite(total)


def _mean(array: ArrayValue) -> Any:
    values = _numbers(array)
    if not values:
        return ABSENT
    total = _sum(array)
    return ABSENT if total is ABSENT else finite(total / len(values))


def _extreme(array: ArrayValue, high: bool) -> Any:
    values = _numbers(array)
    if not values:
        return ABSENT
    best = values[0]
    for one in values:
        if (one > best) if high else (one < best):
            best = one
    return finite(best)


def _spread(array: ArrayValue) -> Any:
    values = _numbers(array)
    mean = _mean(array)
    if values is None or mean is ABSENT:
        return ABSENT
    squares = 0.0
    for one in values:
        deviation = one - mean
        squares += deviation * deviation
    variance = finite(squares / len(values))
    return ABSENT if variance is ABSENT or variance < 0 else finite(math.sqrt(variance))


def _changed(arguments: Sequence[Any], change: Callable[[ArrayValue], Any]) -> Any:
    array = _array(arguments)
    return ABSENT if array is None else change(array)


def _set(array: ArrayValue, arguments: Sequence[Any]) -> Any:
    array.elements[_within(array, arguments[1], len(array.elements))] = arguments[2]
    return ABSENT


def _push(array: ArrayValue, value: Any, front: bool) -> Any:
    if front:
        array.elements.insert(0, value)
    else:
        array.elements.append(value)
    _grew(len(array.elements))
    return ABSENT


def _insert(array: ArrayValue, arguments: Sequence[Any]) -> Any:
    # One past the end is an append: a loop inserting at size(arr) builds a
    # list in order, and refusing it would make the last element a special case.
    array.elements.insert(_within(array, arguments[1], len(array.elements) + 1), arguments[2])
    _grew(len(array.elements))
    return ABSENT


def _take(array: ArrayValue, last: bool) -> Any:
    _empty(array, "the last element" if last else "the first element")
    return array.elements.pop() if last else array.elements.pop(0)


def _remove(array: ArrayValue, arguments: Sequence[Any]) -> Any:
    return array.elements.pop(_within(array, arguments[1], len(array.elements)))


def _slice(arguments: Sequence[Any]) -> Any:
    array = _array(arguments)
    start, stop = arguments[1], arguments[2]
    if array is None or start is ABSENT or stop is ABSENT:
        return ABSENT
    for name, value in (("from", start), ("to", stop)):
        if not is_whole(value) or value < 0:
            raise Refused("OS4003", name="slice", argument=name, found=_spelled(value))
    return _new(array.elements[int(start) : int(stop)])


def _index_of(arguments: Sequence[Any]) -> Any:
    array = _array(arguments)
    if array is None:
        return ABSENT
    for at, one in enumerate(array.elements):
        if equal(one, arguments[1]):
            return float(at)
    return -1.0


def _same(arguments: Sequence[Any]) -> Any:
    a, b = _array(arguments, 0), _array(arguments, 1)
    if a is None or b is None:
        return ABSENT
    if len(a.elements) != len(b.elements):
        return False
    return all(equal(x, y) for x, y in zip(a.elements, b.elements))


def _sort(array: ArrayValue, order: Any) -> Any:
    descending = order == "desc"
    key = functools.cmp_to_key((lambda a, b: _rank(b, a)) if descending else _rank)
    array.elements.sort(key=key)
    return ABSENT


def _reverse(array: ArrayValue) -> Any:
    array.elements.reverse()
    return ABSENT


def _clear(array: ArrayValue) -> Any:
    array.elements.clear()
    return ABSENT


def _reading(of: Callable[[ArrayValue], Any]) -> Callable[[Sequence[Any]], Any]:
    return lambda arguments: _changed(arguments, of)


#: Every array call, by name and argument count, which is how the manifest keys it.
CALLS: Dict[Tuple[str, int], Callable[[Sequence[Any]], Any]] = {
    ("size", 1): _reading(lambda array: float(len(array.elements))),
    ("element", 2): lambda arguments: element(arguments[0], arguments[1])
    if isinstance(arguments[0], ArrayValue)
    else ABSENT,
    ("set", 3): lambda arguments: _changed(arguments, lambda array: _set(array, arguments)),
    ("push", 2): lambda arguments: _changed(arguments, lambda array: _push(array, arguments[1], False)),
    ("unshift", 2): lambda arguments: _changed(arguments, lambda array: _push(array, arguments[1], True)),
    ("insert", 3): lambda arguments: _changed(arguments, lambda array: _insert(array, arguments)),
    ("pop", 1): _reading(lambda array: _take(array, True)),
    ("shift", 1): _reading(lambda array: _take(array, False)),
    ("remove", 2): lambda arguments: _changed(arguments, lambda array: _remove(array, arguments)),
    ("slice", 3): _slice,
    ("copy", 1): _reading(lambda array: _new(list(array.elements))),
    ("indexOf", 2): _index_of,
    ("arrayEqual", 2): _same,
    ("sort", 2): lambda arguments: _changed(arguments, lambda array: _sort(array, arguments[1])),
    ("reverse", 1): _reading(_reverse),
    ("sum", 1): _reading(_sum),
    ("avg", 1): _reading(_mean),
    ("min", 1): _reading(lambda array: _extreme(array, False)),
    ("max", 1): _reading(lambda array: _extreme(array, True)),
    ("stdev", 1): _reading(_spread),
    ("clear", 1): _reading(_clear),
}
